File: Kalorie/Kalorie/Core/UseCases/FetchFoodItemsByIdsUseCase.swift
//
//  FetchFoodItemsByIdsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 19.09.2026.
//

import Foundation

protocol FetchFoodItemsByIdsUseCaseProtocol {
    func callAsFunction(ids: [String]) async throws -> [FoodItemDomain]
}

struct FetchFoodItemsByIdsUseCase: FetchFoodItemsByIdsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol) {
        self.dataProvider = dataProvider
    }

    // MARK: - Functions

    func callAsFunction(ids: [String]) async throws -> [FoodItemDomain] {
        let dtos: [FoodItemDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodItems,
            whereDocumentIdIn: Array(Set(ids))
        )
        return dtos.map { $0.asDomain() }
    }
}

#if DEBUG
struct FetchFoodItemsByIdsUseCaseFake: FetchFoodItemsByIdsUseCaseProtocol {

    // MARK: - Properties

    var stubbedItems: [FoodItemDomain] = []
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(ids: [String]) async throws -> [FoodItemDomain] {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedItems
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchFoodItemsByIdsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

interface FetchFoodItemsByIdsUseCaseProtocol {
    suspend operator fun invoke(ids: List<String>): List<FoodItemDomain>
}

class FetchFoodItemsByIdsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
) : FetchFoodItemsByIdsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(ids: List<String>): List<FoodItemDomain> {
        val dtos: List<FoodItemDTO> = dataProvider.loadAsync(from = Constants.Firestore.FOOD_ITEMS, whereDocumentIdIn = ids.distinct())
        return dtos.map { it.asDomain() }
    }
}
